File: fsharp/simple-cipher/SimpleCipher.fs
module SimpleCipher
open System

let encode (key:string) (input:string) = 
    if Seq.exists (fun (c:char) -> not (Char.IsLetter c) || Char.IsUpper c) key 
        || key.Trim().Length = 0 then failwith "Invalid key"
    let cipher kc ic = 
        let adjust = (int kc) - 97
        let cc = int ic + adjust
        char <| if cc > 122 then cc - 26 else cc
    Seq.map2 cipher key input |> String.Concat

let decode (key:string) (input:string) = 
    let decipher kc cc = 
        let adjust = (int kc) - 97
        let oc = int cc - adjust
        char <| if oc < 97 then oc + 26 else oc
    Seq.map2 decipher key input |> String.Concat

let encodeRandom (input:string) = 
    let random = new Random()
    let key = [0..99] |> Seq.map (fun _ -> random.Next(97,122) |> char) |> String.Concat
    (key, encode key input)

File: csharp/simple-cipher/SimpleCipher.cs
using System;
using System.Linq;

public class SimpleCipher
{
    private readonly int min = (int)'a';
    private readonly int max = (int)'z';
    
    public string Key { get; }

    public SimpleCipher()
    {
        var random = new Random();
        Key = string.Concat(Enumerable.Range(0, 10).Select(_ => (char)random.Next(min, max + 1)));
    }

    public SimpleCipher(string key)
    {
        if(string.IsNullOrWhiteSpace(key) || key.Any(c => !char.IsLetter(c) || !char.IsLower(c)))
            throw new ArgumentException();
        Key = key;
    }

    private int[] KeyForLength(int length) =>
        Enumerable.Range(0, length)
        .Select(i => Key[i % Key.Length] - min)
        .ToArray();

    private char Adjust(char source, int amount)
    {
        var result = source + amount;
        result = result > max ? result - 26 : (result < min ? result + 26 : result);
        return (char)result;
    }

    public string Encode(string plaintext)
    {
        var key = KeyForLength(plaintext.Length);
        return string.Concat(plaintext.Select((c, i) => Adjust(c, key[i])));
    }

    public string Decode(string ciphertext)
    {
        var key = KeyForLength(ciphertext.Length);
        return string.Concat(ciphertext.Select((c, i) => Adjust(c, -key[i])));
    }
}